File: canVisitAllRooms.js
/**
 * @param {number[][]} rooms
 * @return {boolean}
 */
var canVisitAllRooms = function (rooms) {
  let seen = Array(rooms.length);
  let stack = [0];
  while (stack.length) {
    let cur = stack.pop();
    if (!seen[cur]) {
      seen[cur] = 1;
      for (let k of rooms[cur]) {
        if (!seen[k]) stack.push(k);
      }
    }
  }
  for (let r of seen) {
    if (!r) return false;
  }
  return true;
};
var canVisitAllRooms = function (rooms) {
  let stack = [0];
  let seen = new Set(stack);

  while (stack.length) {
    let cur = stack.pop();
    for (let k of rooms[cur]) {
      if (!seen.has(k)) {
        stack.push(k);
        seen.add(k);
        if (seen.size === rooms.length) return true;
      }
    }
  }
  return seen.size === rooms.length;
};
canVisitAllRooms([[1], [2], [3], []]);
canVisitAllRooms([[4], [3], [], [2, 5, 7], [1], [], [8, 9], [], [], [6]]);

File: leetcode/canVisitAllRooms.py
class Solution:
    def canVisitAllRooms(self, rooms) -> bool:
        stack = [0]
        seen = set(stack)
        while stack:
            cur = stack.pop()
            for k in rooms[cur]:
                if k not in seen:
                    stack.append(k)
                    seen.add(k)
        return len(seen) == len(rooms)


def can_visit_2(rooms):
    n = len(rooms)
    seen = [True] + [False] * (n-1)
    stack = [0]
    while stack:
        cur = stack.pop()
        for k in rooms[cur]:
            if not seen[k]:
                stack.append(k)
                seen[k] = True
    return all(seen)


def can_visit(rooms):
    stack = [0]
    seen = set()
    while stack:
        cur = stack.pop()
        seen.add(cur)
        for k in rooms[cur]:
            if k not in seen:
                stack.append(k)
    return len(seen) == len(rooms)


Solution().canVisitAllRooms([[1], [2], [3], []])
